cobaya/theory.py

```diff
@@ -75,21 +75,64 @@ def get_requirements(self):
         :return: dictionary of requirements (or iterable of requirement names if no
                  optional parameters are needed)
         """
-
         return dict.fromkeys(str_to_list(getattr(self, _requires, [])))
 
-    def needs(self, **requirements):
+    def _update_required_params(self):
+        """
+        Selects from the list of requirements those that appear to be parameters (i.e.
+        they are indicated in the requirements with no options *and* are handed by the
+        provider through the set_params method.
+
+        Before the call to `initialize_with_provider`, only the first condition is
+        checked.
+        """
+        self._required_params = set()
+        reqs = self.get_requirements()
+        if not isinstance(reqs, Mapping):
+            reqs = dict.fromkeys(reqs)
+        for req, opt in reqs.items():
+            if opt is None:
+                if getattr(self, "provider", None) is not None:
+                    if req in self.provider.requirement_providers:
+                        this_provider = self.provider.requirement_providers[req]
+                        if req in this_provider.get_can_provide_params():
+                            self._required_params.add(req)
```

Comment:
This peeking inside the implementation of Provider (an indirectly using properties of other theories) is a bit odd, and not really in the spirit of class separation.

[without this, even the existence of self.provider is an implementation choice rather than enforced - just the default initialize_with_params() implementation chooses to store the Provider instance into self.provider as a default convention]

Reply:
This is not done any more (this method has been removed), since it is actually not necessary at least for the cases we work with right now to distinguish between parameters and other requirements.